core/tee/tee_rpmb_fs.c

```diff
@@ -1472,54 +1514,364 @@ static TEE_Result tee_rpmb_get_max_block(uint16_t dev_id, uint32_t *max_block)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct rpmb_fat_entry *fe = NULL;
+	uint32_t num_elems_read;
+	uint32_t fat_address_local;
+
+	/* Must call fat_entry_dir_init first. */
+	if (!fat_entry_dir || !fat_entry)
+		goto out;
+
+	/* Don't read further if we previously read the last FAT FS entry. */
+	if (fat_entry_dir->last_reached) {
+		res = TEE_SUCCESS;
+		*fat_entry = NULL;
+		goto out;
+	}
+
+	fe = fat_entry_dir->rpmb_fat_entry_buf;
+
+	/* Determine address of FAT FS entry in RPMB storage. */
+	fat_address_local = RPMB_FS_FAT_START_ADDRESS +
+		(fat_entry_dir->num_total_read *
+			sizeof(struct rpmb_fat_entry));
+
+	/*
+	 * We've read all so-far buffered elements, so we need to
+	 * read in more entries from RPMB storage.
+	 */
+	if (fat_entry_dir->idx_curr >= fat_entry_dir->num_buffered) {
+		/*
+		 * This is the case where we do not cache entries, so just read
+		 * in next set of FAT fs entries into the buffer.
+		 * Goto the end of the when statement if that is done.
+		 */
+		if (CFG_RPMB_FS_CACHE_SIZE == 0) {
+			num_elems_read = CFG_RPMB_FS_RD_ENTRIES;
+			fat_entry_dir->idx_curr = 0;
+
+			res = tee_rpmb_read(CFG_RPMB_FS_DEV_ID,
+					    fat_address_local, (uint8_t *)fe,
+					    num_elems_read * sizeof(*fe), NULL,
+					    NULL);
+
+			if (res != TEE_SUCCESS)
+				goto out;
+			else
+				goto post_read_in;
+		}
+
+		/*
+		 * We cache FAT FS entries, and the buffer is not completely
+		 * filled. Further keep on extending the buffer up to its max
+		 * size by reading in from RPMB.
+		 */
+		if (fat_entry_dir->num_total_read < RPMB_BUF_MAX_SIZE) {
+			/*
+			 * Read at most as many elements as fit in the buffer
+			 * and no more than the defined number of entries to
+			 * read in at once.
+			 */
+			num_elems_read = MIN(RPMB_BUF_MAX_SIZE -
+				fat_entry_dir->num_total_read,
+				(uint32_t)CFG_RPMB_FS_RD_ENTRIES);
+
+			/*
+			 * Expand the buffer to fit in the additional entries.
+			 */
+			fe = realloc(fe,
+				     (fat_entry_dir->num_buffered +
+				     num_elems_read) * sizeof(*fe));
+			if (!fe) {
+				res = TEE_ERROR_OUT_OF_MEMORY;
+				goto out;
 			}
 
-			/* Move to next fat_entry. */
-			fat_address += sizeof(struct rpmb_fat_entry);
+			fat_entry_dir->rpmb_fat_entry_buf = fe;
+
+			/* Read in to the next free slot in the buffer/cache. */
+			res = tee_rpmb_read(CFG_RPMB_FS_DEV_ID,
+					    fat_address_local,
+					    (uint8_t *)(fe +
+					    fat_entry_dir->num_total_read),
+					    num_elems_read * sizeof(*fe),
+					    NULL, NULL);
+			if (res != TEE_SUCCESS)
+				goto out;
+
+			/* Update number of buffered elements. */
+			fat_entry_dir->num_buffered += num_elems_read;
+		/*
+		 * This happens when we have read as many elements as can
+		 * possibly fit into the buffer.
+		 * As the first part of the buffer serves as our cache,
+		 * we only overwrite the last part that serves as our temporary
+		 * buffer used to iteratively read in entries when the cache
+		 * is full.
+		 */
```

Comment:
This comment is not at the right place, please move it down inside the `else` block (and with the proper indentation).